README.md
<h1 align="center">Last.fm Stats</h1>

<div align="center">
  <strong>Visualize all your Last.fm activity in a beautiful and interactive way.</strong>

  ![Travis (.org)](https://img.shields.io/travis/davidomarf/lastfm.svg?style=flat-square)
  ![Code Climate maintainability](https://img.shields.io/codeclimate/maintainability/davidomarf/lastfm.svg?style=flat-square)
  ![Codacy grade](https://img.shields.io/codacy/grade/a4df9a5dc6a44d1c901787988a0fc6c2.svg?style=flat-square)
  <br/>
  
</div>

<p align="center">
  <img src="./img/lastfm-screenshot.png" alt="LastfmStats screenshot displaying davidomarf yearly heatmap">
  <a href="https://lastfmstats.herokuapp.com/">
  <i>Try Yours </i>
  </a>
</p>


## Table of Contents

- [Features](#features)
- [Tools](#tools)
- [Installation](#installation)
- [License](#license)

## Features

**bold features** are implemented, `em features` are a work in progress,
_italic features_ are planned, but not yet in development.

### Data Visualization

- **yearly heatmap:** that shows the most and least active days, over a yearly period.
  Measured by the number of scrobbles.
  It defaults to the period of _[1 year ago, today]_.

- `yearly time series:` that shows the number of scrobbles per week, over a yearly period.
  For every week, it shows the song, artist, and album most listened to.
  It defaults to the period of _[1 year ago, today]_.

- _yearly artist time series:_ that shows the number of scrobbles from a single artist
  per week, over a yearly period.
  For every week, it shows the most listened to song.
  _Allow to switch the measurement method between scrobbles and listening time._

- _overall decades histogram:_ that shows the number of `[scrobbles, unique tracks]` for
  songs **originally** released by decade.

- _language distribution (pie?):_ that shows the languages proportion for the `[scrobbles, unique tracks]`
  in the library.

- _explicit vs. non-explicit:_ that shows the percentage of `[scrobbles, unique tracks]` with
  explicit content.

### Playlist creation

- _The `n` most listened to `[songs, artists, albums]` for every `[day, week, month, year]`,
  over `[week, month, year, overal, custom]`._

## Tools

This project is being built using [React](https://reactjs.org/). 

To display page sections as a fullscreen element, we use
[fullPage.js](https://github.com/alvarotrigo/fullpage.js/)

For the visualizations, we use [d3.js](https://d3js.org/).

## Installation

To use the tool, you just need to visit https://lastfmstats.herokuapp.com/

To contribute with the development you'll need a working copy of the repo in your working machine.

- You'll need to have `npm` or `yarn` installed.

- Fork the repository.

- Clone and open the directory
  ```sh
  $ git clone git@github.com:{user}/lastfm.git
  $         # https://github.com/{user}/lastfm.git
  $ cd lastfm
  ```

- [Get an API Key from last.fm](https://www.last.fm/api/account/create)

- Create a `.env` file with this content: `REACT_APP_LAST_FM_API={your-api}`

- Available scripts

  - `npm start` / `yarn start`
    -  Runs the app in the development mode. Open http://localhost:3000 to view it in the browser.
  - `npm test` / `yarn test`
    - Launches the test runner in the interactive watch mode.
  - `npm run build` / `yarn run build`
    - Builds the app for production to the build folder.

## License

[MIT](https://tldrlegal.com/license/mit-license)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Changes that are *meta*, like updating community files, changing the readme, or
modifying templates, will be under the heading *Meta*. These changes don't affect 
the app functionality.

## [Unreleased]

## [0.2.0] - 2019-06-04
### Added
- Loading page to indicate the progress in retreiving the user scrobbles
- Time Series component that plots the weekly scrobbles over the last year
### Changed
- Stop drawing plots incrementally. Fetch first, plot later.

## [0.1.3] - 2019-06-29
### Fixed
- Fix the imports in UserPage replacing import with require

## [0.1.2] - 2019-06-29
### Fixed
- Specify node version in package.json to fix Heroku deployment

## [0.1.1] - 2019-06-29
### Changed
- Draw Heatmap incrementally, updating every time UserPage fetches another page of
scrobbles
- The Heatmap cells have an id based on the date they display. Previously, it was
based on the days elapsed since the start of the displayed period.
- Change the color palete of the Heatmap. 
- Reduce the number of color tags from 8 to 4.

### Fixed
- Tooltip no longer goes offscreen but instead now flips to the left

## [0.1.0] - 2019-06-26
### Added
- Create a React project
- Manage routes `/user/:user`
- Create a form to ask for the username to fetch
- Get a list of all the scrobbles from the user in `user/:user`
- Display a Heatmap of the last year (Today but one year ago, today)

#### Meta
- Create a basic GitHub repository
- Add community files: templates, contributing, code of conduct.
- Write a useful Readme

[Unreleased]: https://github.com/davidomarf/lastfm/compare/v0.2.0..develop
[0.2.0]: https://github.com/davidomarf/lastfm/compare/tag/v0.2.0
[0.1.3]: https://github.com/davidomarf/lastfm/compare/tag/v0.1.3
[0.1.2]: https://github.com/davidomarf/lastfm/compare/tag/v0.1.2
[0.1.1]: https://github.com/davidomarf/lastfm/compare/tag/v0.1.1
[0.1.0]: https://github.com/davidomarf/lastfm/releases/tag/v0.1.0


src/utils/visualization-utils.test.js
const utils = require("./visualization-utils");

describe("Visualization utils", () => {
  describe("getIDFromDate ()", () => {
    it("should set a default prefix", () => {
      expect(utils.getIDFromDate(new Date(2000, 5, 10))).toBe("id-2000-6-10");
    });

    it("should use a sent prefix", () => {
      expect(utils.getIDFromDate(new Date(1998, 10, 10), "date")).toBe(
        "date-1998-11-10"
      );
    });

    it("should fail with non-date objects", () => {
      expect(() => utils.getIDFromDate("date", "pre")).toThrow();
    });
  });

  describe("getDateArray ()", () => {
    it("should have n+1 dates in the array when dates are n days apart", () => {
      expect(
        utils.getDateArray(
          new Date(),
          new Date(new Date().setDate(new Date().getDate() + 10))
        )
      ).toHaveLength(11);
    });

    it("should return one date if dates are the same", () => {
      expect(utils.getDateArray(new Date(), new Date())).toHaveLength(1);
    });

    it("should throw when start is bigger than end", () => {
      expect(() =>
        utils.getDateArray(
          new Date(new Date().setDate(new Date().getDate() + 10)),
          new Date()
        )
      ).toThrow();
    });
  });
});


src/API/lastfm.js
const axios = require("axios");

const lastFmAPI = process.env.REACT_APP_LAST_FM_API;
const lastFmCall =
  "https://ws.audioscrobbler.com/2.0/?format=json&api_key=" + lastFmAPI;

async function getData(url) {
  try {
    const response = await axios.get(url);
    return response.data;
  } catch (error) {
    console.log(error);
  }
}

// Exported functions used in UserPage

/**
 * Returns the number of scrobbles the user has
 * @param {string} user Last.fm user
 * @returns {Promise} Response to the method https://www.last.fm/api/show/user.getInfo
 */
function getUserPlayCount(user) {
  let playCount = getData(lastFmCall + "&method=user.getinfo&user=" + user);
  return playCount;
}

/**
 * Returns the specified page from the user's scrobble list. Each page has 200 entries.
 * @param {string} user
 * @param {number} page Page of the
 * @returns {Promise} Response to the method https://www.last.fm/api/show/user.getRecentTracks
 */
function getScrobbles(user, page) {
  return getData(
    `${lastFmCall}&method=user.getrecenttracks&limit=200&user=${user}&page=${page}`
  );
}

module.exports = {
  getScrobbles,
  getUserPlayCount
};


src/Visualizations/TimeSeries.js
/**
 * Time series graph that plots the weekly scrobbles in a year.
 *
 * The data is received as a prop from the state of UserPage, it calculates the
 * number of scrobbles on each week, and plots week vs. scrobbles in a time
 * series graph.
 *
 * @file    This file defines the TimeSeries Component.
 * @author  davidomarf
 * @since   v0.2.0
 */

import React from "react";
import styles from "./TimeSeries.module.scss";
import * as d3 from "d3";
import * as utils from "../utils/visualization-utils";

const lineFunction = d3
  .line()
  .x((d) => d3.select("#" + d.key).attr("cx"))
  .y((d) => d3.select("#" + d.key).attr("curve-y"))
  .curve(d3.curveMonotoneX);

/* -------------------------- Component Definition -------------------------- */

/**
 * TimeSeries that plots the weekly scrobbles in a year.
 */
class TimeSeries extends React.Component {
  constructor(props) {
    super(props);

    // List containing {id:count} elements, where id is the ID generated using
    // getIDFromDate(), and count is the number of scrobbles in that date
    this.frequencyList = {};

    // Maximum number of scrobbles in a single day
    this.mostScrobblesInAWeek = 0;

    // The last date for the period to be displayed in the Heatmap
    this.endDate = { date: new Date() };
    this.endDate.uts = Math.ceil(this.endDate.date.getTime() / 1000);

    // Same day as endDate, but one year ago
    this.startDate = {
      date: new Date(
        new Date().setFullYear(this.endDate.date.getFullYear() - 1)
      )
    };
    this.startDate.utc = Math.floor(this.startDate.date.getTime() / 1000);
  }

  /**
   * Update the Heatmap to indicate the number of scrobbles in each day.
   *
   * Go through an "array-ized" version of frequencyList, and use the {id} to
   * select the SVG items, and the {value} to calculate the fill color.
   */
  updateHeatmapValues() {
    // Convert the dictionary into an array of {key, value} objects
    // @see https://github.com/d3/d3-collection#entries
    let freqArray = d3.entries(this.frequencyList);
    this.numberOfColorTags = 4;

    // Update using the maximum value in freqArray
    this.mostScrobblesInAWeek = d3.max(freqArray, (d) => d.value);

    for (let i = 0; i < freqArray.length; i++) {
      d3.select("#" + freqArray[`${i}`].key)
        .attr(
          "curve-y",
          -(freqArray[`${i}`].value / this.mostScrobblesInAWeek) * 60
        )
        .attr("scrobbles", freqArray[`${i}`].value)
        .transition()
        .duration(2000)
        .attr(
          "cy",
          -(freqArray[`${i}`].value / this.mostScrobblesInAWeek) * 60
        );
      d3.select("#ts-curve")
        .transition()
        .duration(2000)
        .attr("d", lineFunction(freqArray));
    }
  }

  drawHeatmapCells(svg) {
    let firstMonday = new Date(this.startDate.date).setDate(
      this.startDate.date.getDate() - this.startDate.date.getDay()
    );

    let dates = utils.getDateArray(firstMonday, this.endDate.date, 7);
    let style = styles["week-circle"];

    svg
      .append("path")
      .attr("id", "ts-curve")
      .attr("d", lineFunction(d3.entries(this.frequencyList)))
      .attr("stroke", "#4e5467")
      .attr("stroke-width", 0.5)
      .attr("fill", "none");
    let monthShift = 0;
    for (let i = 0; i < dates.length; i++) {
      let id = utils.getIDFromDate(dates[`${i}`], "ts");
      this.frequencyList[`${id}`] = 0;
      if (dates[`${i}`].getDate() <= 7) {
        monthShift += 5;
      }
      svg
        .append("circle")
        .attr("cx", 22 + monthShift+ i * 10)
        .attr("curve-y", 0)
        .attr("cy", 0)
        .attr("r", 1)
        .attr("class", style)
        .attr("id", id)
        // The attributes 'date' and 'scrobbles' are used to get the content
        // of the tooltip when hovering on a cell
        .attr("date", utils.formatDate(dates[`${i}`]))
        .attr("scrobbles", 0)
        .on("mouseover", mouseover)
        .on("mousemove", mousemove)
        .on("mouseleave", mouseleave);
    }
  }

  /**
   * React function to run when the component updates
   *
   * It's expected that the component will be updated every time UserPage
   * fetches one more page of the scrobbles list.
   */
  componentDidUpdate() {}

  /**
   * React function to run when the component has been mounted
   *
   * This runs only once, and uses that fact to initialize the canvas, and append
   * the items that won't be redrawn.
   */
  componentDidMount() {
    // This creates the svg for the graph
    const svg = d3
      .select(`#d3-section-${this.props.title}`)
      .append("svg")
      // This makes the svg element responsive
      // @see https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/preserveAspectRatio
      // @see https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/viewBox
      .attr("preserveAspectRatio", "xMinYMin meet")
      .attr("viewBox", "-20 -119 640 120");
    // .attr("viewBox", "-20 -250 560 270");

    // // Draw the Heatmap structure: Legend, texts, and cells
    // this.drawLegend(svg);
    // this.drawWeekDays(svg);
    this.drawHeatmapCells(svg);

    // Initialize the tooltip. Draw it after all other items so it stays at top
    tooltip = svg.append("rect");
    tooltip_text = svg.append("text");

    // Assign the props value to a variable to avoid writing a lot
    let lastList = this.props.user.scrobbles;

    // Convert the list of scrobble objects into a list of strings using the
    // date to determine an ID
    let idList = [].concat.apply(
      [],
      lastList.map((e) =>
        e.list
          .map((e) => {
            let weekMonday = new Date(1000 * Math.ceil(Number(e.date.uts)));
            weekMonday.setDate(weekMonday.getDate() - weekMonday.getDay());
            return utils.getIDFromDate(weekMonday, "ts");
          })
          .filter((e) => this.frequencyList[`${e}`] !== undefined)
      )
    );

    // This counts the number of times that a given id appears in the
    // current list of ids
    for (let i = 0; i < idList.length; i++) {
      let id = idList[`${i}`];
      // If this.frequencyList.id is defined, increment its value
      if (this.frequencyList[`${id}`] !== undefined) {
        this.frequencyList[`${id}`]++;
      }
    }

    // Update the values considering the new added scrobbles
    this.updateHeatmapValues();
  }

  render() {
    return (
      <>
        <div
          id={`d3-section-${this.props.title}`}
          className={styles["graph-container"]}
        />
      </>
    );
  }
}

/* --------------------------- D3 Tooltip Elements -------------------------- */

// Tooltip and Tooltip text are global variables that are used to display the
// number of scrobbles and date of eaach day in the Heatmap.
let tooltip;
let tooltip_text;

// Rules to modify the tooltip the mouse is over
let mouseover = function(d) {
  tooltip.style("opacity", 0.9);
  tooltip_text.style("opacity", 1);
};

// Rules to modify the tooltip when the mouse moves within it
let mousemove = function(d) {
  let week = d3.select(this);
  tooltip_text
    .style("fill", "white")
    .text(`${week.attr("scrobbles")} scrobbles on ${week.attr("date")}`)
    .attr("class", styles["meta-text"])
    .attr(
      "x",
      d3.mouse(this)[0] +
        (d3.mouse(this)[0] > 410
          ? -(tooltip_text.node().getBBox().width + 15)
          : 15)
    )
    .attr("y", d3.mouse(this)[1]);
  tooltip
    .attr("width", tooltip_text.node().getBBox().width + 10)
    .attr("height", 15)
    .attr(
      "x",
      d3.mouse(this)[0] +
        (d3.mouse(this)[0] > 410
          ? -(tooltip_text.node().getBBox().width + 20)
          : 10)
    )
    .attr("y", d3.mouse(this)[1] - 10);
};

// Rules to modify the tooltip when the mouse leaves it
let mouseleave = function(d) {
  tooltip.style("opacity", 0);
  tooltip_text.text("");
};

/* ---------------------------- Export Component ---------------------------- */

export default TimeSeries;


src/utils/visualization-utils.js
/**
 * Utilities to unclutter the visualization components.
 *
 * @file    This file defines helping functions for the visualization components.
 * @author  davidomarf
 * @since   v2.0.0
 */

// Used to render the month's name without calling Date.toLocaleDateString()
export const months = [
  "Jan",
  "Feb",
  "Mar",
  "Apr",
  "May",
  "Jun",
  "Jul",
  "Aug",
  "Sep",
  "Oct",
  "Nov",
  "Dec"
];

/**
 * Formats a Date in a readable String.
 *
 * The format is Month dd, YYYY.
 *
 * @param {Object} date  Date to be formatted
 * @returns {string}     Date to string in format Month dd, YYYY
 */
export function formatDate(date) {
  return `${months[date.getMonth()]} ${date.getDate()}, ${date.getFullYear()}`;
}

/**
 * Generate an array of dates containing every day between start and end.
 *
 * You must ensure start is lower than end
 *
 * @param {Date} start      First value of the array
 * @param {Date} end        Last value of the array
 * @param {number} [step=1] Step in days between the dates in the array
 * @return {Date[]}         All the dates between start and end
 */
export function getDateArray(start, end, step = 1) {
  if (start > end) {
    throw new Error(
      "Start date is bigger than end date (start is in the future of end)"
    );
  }
  let arr = [];
  let dt = new Date(start);

  while (dt <= end) {
    arr.push(new Date(dt));
    dt.setDate(dt.getDate() + step);
  }

  return arr;
}

/**
 * Creates an ID for a date using its Year, Month, Date.
 *
 * It appends a prefix that should be unique for every instance the function is called
 * to distinguish between different elements that create elements using the same date.
 *
 * @param {Date} date           Date to be used
 * @param {string} [prefix="id"]  Prefix to append before the readable date
 * @returns {string}              prefix + Date in format YYYY-MM-D
 */
export function getIDFromDate(date, prefix = "id") {
  return `${prefix}-${date.getFullYear()}-${date.getMonth() +
    1}-${date.getDate()}`;
}


src/Visualizations/Heatmap.js
/**
 * Heatmap that plots the daily scrobbles in a year.
 *
 * The data is received as a prop from the state of UserPage, it calculates the
 * number of scrobbles on each day, and assigns each day a color based on the
 * number of scrobbles and the maximum number of daily scrobbles, so days close
 * to that number are brighter, and days close to 0, are darker.
 *
 * @file    This file defines the Heatmap Component.
 * @author  davidomarf
 * @author  JAKESNAKE523
 * @since   v0.1.0
 */

import React from "react";
import styles from "./Heatmap.module.scss";
import * as d3 from "d3";
import * as utils from "../utils/visualization-utils";

/* -------------------------- Component Definition -------------------------- */

/**
 * Heatmap that plots the daily scrobbles in a year.
 */
class Heatmap extends React.Component {
  constructor(props) {
    super(props);
    // List containing {id:count} elements, where id is the ID generated using
    // utils.getIDFromDate(), and count is the number of scrobbles in that date
    this.frequencyList = {};

    // Maximum number of scrobbles in a single day
    this.mostScrobblesInADay = 0;

    // Doesn't include the color assigned to days with 0 scrobbles.
    // If needs update, also update myColor...domain([0, {newValue}])
    this.numberOfColorTags = 4;

    // The last date for the period to be displayed in the Heatmap
    this.endDate = { date: new Date() };
    this.endDate.utc = Math.ceil(this.endDate.date.getTime() / 1000);

    // Same day as endDate, but one year ago
    this.startDate = {
      date: new Date(
        new Date().setFullYear(this.endDate.date.getFullYear() - 1)
      )
    };
    this.startDate.utc = Math.floor(this.startDate.date.getTime() / 1000);
  }

  /**
   * Writes the name of the days for the rows.
   *
   * Write "Mon", "Wed", and "Fri" at the leftmost part of the visualization
   * for the corresponding rows.
   */
  drawWeekDays() {
    let days = ["Mon", "Wed", "Fri"];
    for (let i = 0; i < days.length; i++) {
      this.svg
        .append("text")
        .attr("x", 10)
        .attr("y", (i + 1) * 20 - 4.25)
        .attr("class", styles["meta-text"] + " " + styles["day-flag"])
        .text(days[`${i}`]);
    }
  }

  /**
   * Draws the legend of the graph.
   *
   * The legend will display all the available color tags, and the range of
   * values each one indicates.
   */
  drawLegend() {
    let text = [
      [20, 100, "Less"],
      [45 + (this.numberOfColorTags + 1) * 10, 100, "More"]
    ];

    this.svg
      .selectAll("text")
      .data(text)
      .enter()
      .append("text")
      .attr("x", (d) => d[0])
      .attr("y", (d) => d[1])
      .attr("class", styles["meta-text"])
      .text((d) => d[2]);

    for (let i = 0; i <= this.numberOfColorTags; i++) {
      this.svg
        .append("rect")
        .attr("x", 42 + i * 10)
        .attr("y", 92)
        .attr("width", 9)
        .attr("height", 9)
        .attr("style", `fill:${myColor(i)}; stroke:#4E5467`);
    }
  }

  /**
   * Update the Heatmap to indicate the number of scrobbles in each day.
   *
   * Go through an "array-ized" version of frequencyList, and use the {id} to
   * select the SVG items, and the {value} to calculate the fill color.
   */
  updateHeatmapValues() {
    // Convert the dictionary into an array of {key, value} objects
    // @see https://github.com/d3/d3-collection#entries
    let freqArray = d3.entries(this.frequencyList);

    // Update using the maximum value in freqArray
    this.mostScrobblesInADay = d3.max(freqArray, (d) => d.value);

    for (let i = 0; i < freqArray.length; i++) {
      // The number sent to myColor is an integer between 0 and
      // this.numberOfColorTags. It is calculated by multiplying the ratio of
      // dayValue / maxValue by the number of available color tags.
      let color = myColor(
        Math.ceil(
          this.numberOfColorTags *
            (freqArray[`${i}`].value / this.mostScrobblesInADay)
        )
      );

      d3.select("#" + freqArray[`${i}`].key)
        .attr("scrobbles", freqArray[`${i}`].value)
        .transition()
        .duration(2000)
        .attr(
          "r",
          1.5 +
            this.numberOfColorTags *
              0.5 *
              (freqArray[`${i}`].value / this.mostScrobblesInADay)
        )
        .style("fill", color)
        .style("stroke", color);
    }
  }

  drawHeatmapCells() {
    // Create an array of dates for the year
    let dates = utils.getDateArray(this.startDate.date, this.endDate.date);

    // The shifting values are the ones that displace the cells in the x
    // coordinate every new week or month
    let monthShift = 0;
    let weekShift = 0;

    // When true, will write the name of the month over the first sunday
    // it encounters
    let writeMonthNameNextSunday = false;

    for (let i = 0; i < dates.length; i++) {
      let style = styles["day"];

      // .getDay() returns the day of the week (between 0, 6)
      // .getDate() returns the day of the month
      let weekDay = dates[`${i}`].getDay();
      let monthDay = dates[`${i}`].getDate();

      // This makes the script write the name of the month on the second week
      // of the month
      if (monthDay === 1) {
        writeMonthNameNextSunday = true;
        // Also increment the shift for the cells of the next month. This
        // creates a space between months.
        monthShift += 5;
      }

      // When it's sunday
      if (weekDay === 0) {
        // When starting a new week, increment the shift for the new cells
        weekShift++;

        // If we have to write the name of the month on top
        if (writeMonthNameNextSunday) {
          this.svg
            .append("text")
            .attr("x", 20 + monthShift + weekShift * 10)
            .attr("y", 80)
            .text(utils.months[dates[`${i}`].getMonth()])
            .attr("class", styles["meta-text"] + " " + styles["month-flag"]);

          // Avoid drawing the month more than once
          writeMonthNameNextSunday = false;
        }
      }

      /* ---------------------- Append a heatmap cell --------------------- */

      // Get the ID for the cell using the date
      let id = utils.getIDFromDate(dates[`${i}`], "hm");

      this.svg
        .append("circle")
        .attr("cx", 22 + monthShift + weekShift * 10)
        .attr("cy", 2 + weekDay * 10)
        .attr("r", 1)
        // .attr("height", 8)
        .attr("id", id)
        .attr("class", style)
        // The attributes 'date' and 'scrobbles' are used to get the content
        // of the tooltip when hovering on a cell
        .attr("date", utils.formatDate(dates[`${i}`]))
        .attr("scrobbles", 0)
        .on("mouseover", mouseover)
        .on("mousemove", mousemove)
        .on("mouseleave", mouseleave);
    }
  }

  /**
   * React function to run when the component has been mounted
   *
   * This runs only once, and uses that fact to initialize the canvas, and append
   * the items that won't be redrawn.
   */
  componentDidMount() {
    // Draw the Heatmap structure: Legend, texts, and cells
    // This creates the svg for the graph
    this.svg = d3
      .select(`#d3-section-${this.props.title}`)
      .append("svg")
      // This makes the svg element responsive
      // @see https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/preserveAspectRatio
      // @see https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/viewBox
      .attr("preserveAspectRatio", "xMinYMin meet")
      .attr("viewBox", "-20 -5 640 120");

    this.drawLegend();
    this.drawWeekDays();
    this.drawHeatmapCells();

    // Initialize the tooltip. Draw it after all other items so it stays at top
    tooltip = this.svg.append("rect");
    tooltip_text = this.svg.append("text");

    this.updateFrequencyList(this.props.user.scrobbles);

    // Update the values considering the new added scrobbles
    this.updateHeatmapValues();
  }

  updateFrequencyList(scrobbleList) {
    // Convert the list of scrobble objects into a list of strings using the
    // date to determine an ID
    let idList = [].concat.apply(
      [],
      scrobbleList.map((e) =>
        e.list.map((e) =>
          utils.getIDFromDate(new Date(1000 * Number(e.date.uts)), "hm")
        )
      )
    );
    
    // This counts the number of times that a given id appears in the
    // current list of ids
    for (let i = 0; i < idList.length; i++) {
      let id = idList[`${i}`];
      // If this.frequencyList.id is defined, increment its value
      if (this.frequencyList[`${id}`]) {
        this.frequencyList[`${id}`]++;
      }
      // If not, it means it's the first time that id appears, so initialize it
      else {
        this.frequencyList[`${id}`] = 1;
      }
    }
  }

  render() {
    return (
      <>
        <div
          id={`d3-section-${this.props.title}`}
          className={styles["graph-container"]}
        />
      </>
    );
  }
}

/**
 * Function to interpolate colors using a given d3 color palette.
 *
 * The domain is the range of the expected values to be received in the
 * interpolator. For this use case, domain[1] must match
 * Heatmap.numberofCategories.
 */
const myColor = d3
  .scaleLinear()
  .range(["#383C4A", "#e4fd3f"])
  .domain([0, 4]);

/* --------------------------- D3 Tooltip Elements -------------------------- */

// Tooltip and Tooltip text are global variables that are used to display the
// number of scrobbles and date of each day in the Heatmap.
let tooltip;
let tooltip_text;

// Rules to modify the tooltip the mouse is over
let mouseover = function(d) {
  tooltip.style("opacity", 0.9);
  tooltip_text.style("opacity", 1);
};

// Rules to modify the tooltip when the mouse moves within it
let mousemove = function(d) {
  let day = d3.select(this);
  tooltip_text
    .style("fill", "white")
    .text(`${day.attr("scrobbles")} scrobbles on ${day.attr("date")}`)
    .attr("class", styles["meta-text"])
    .attr(
      "x",
      d3.mouse(this)[0] +
        (d3.mouse(this)[0] > 491
          ? -(tooltip_text.node().getBBox().width + 15)
          : 15)
    )
    .attr("y", d3.mouse(this)[1]);
  tooltip
    .attr("width", tooltip_text.node().getBBox().width + 10)
    .attr("height", 15)
    .attr(
      "x",
      d3.mouse(this)[0] +
        (d3.mouse(this)[0] > 491
          ? -(tooltip_text.node().getBBox().width + 20)
          : 10)
    )
    .attr("y", d3.mouse(this)[1] - 10);
};

// Rules to modify the tooltip when the mouse leaves it
let mouseleave = function(d) {
  tooltip.style("opacity", 0);
  tooltip_text.text("");
};

/* ---------------------------- Export Component ---------------------------- */

export default Heatmap;
